src/protocols/der/der.h

```diff
@@ -0,0 +1,171 @@
+#include "include/build.h"
+#include "lib/util/types.h"
+#include <freeradius-devel/util/dict.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+extern HIDDEN fr_dict_t const *dict_der;
+
+/** Enumeration describing the data types in a DER encoded structure
+ */
+typedef enum {
+	FR_DER_TAG_BOOLEAN	    = 0x01,	   //!< Boolean true/false
+	FR_DER_TAG_INTEGER	    = 0x02,	   //!< Arbitrary width signed integer.
+	FR_DER_TAG_BITSTRING	    = 0x03,	   //!< String of bits (length field specifies bits).
+	FR_DER_TAG_OCTETSTRING	    = 0x04,	   //!< String of octets (length field specifies bytes).
+	FR_DER_TAG_NULL		    = 0x05,	   //!< An empty value.
+	FR_DER_TAG_OID		    = 0x06,	   //!< Reference to an OID based attribute.
+	FR_DER_TAG_ENUMERATED	    = 0x0a,	   //!< An enumerated value.
+	FR_DER_TAG_UTF8_STRING	    = 0x0c,	   //!< String of UTF8 chars.
+	FR_DER_TAG_SEQUENCE	    = 0x10,	   //!< A sequence of DER encoded data (a structure).
+	FR_DER_TAG_SET		    = 0x11,	   //!< A set of DER encoded data (a structure).
+	FR_DER_TAG_PRINTABLE_STRING = 0x13,	   //!< String of printable chars.
+	FR_DER_TAG_T61_STRING	    = 0x14,	   //!< String of T61 (8bit) chars.
+	FR_DER_TAG_IA5_STRING	    = 0x16,	   //!< String of IA5 (7bit) chars.
+	FR_DER_TAG_UTC_TIME	    = 0x17,	   //!< A time in UTC "YYMMDDhhmmssZ" format.
+	FR_DER_TAG_GENERALIZED_TIME = 0x18,	   //!< A time in "YYYYMMDDHHMMSS[.fff]Z" format.
+	FR_DER_TAG_VISIBLE_STRING   = 0x1a,	   //!< String of visible chars.
+	FR_DER_TAG_GENERAL_STRING   = 0x1b,	   //!< String of general chars.
+	FR_DER_TAG_UNIVERSAL_STRING = 0x1c,	   //!< String of universal chars.
+	FR_DER_TAG_BMP_STRING	    = 0x1e,	  //!< String of BMP chars.
+
+	FR_DER_TAG_CHOICE	    = 0x23,	   //!< A choice of types. Techically not a DER tag, but used to represent a choice.
+
+	FR_DER_TAG_MAX		= UINT8_MAX
+} fr_der_tag_num_t;
+
+typedef enum {
+	FR_DER_TAG_PRIMATIVE   = 0x00,	      //!< This is a leaf value, it contains no children.
+	FR_DER_TAG_CONSTRUCTED = 0x20	     //!< This is a sequence or set, it contains children.
+} fr_der_tag_constructed_t;
+
+typedef enum {
+	FR_DER_CLASS_UNIVERSAL   = 0x00,
+	FR_DER_CLASS_APPLICATION = 0x40,
+	FR_DER_CLASS_CONTEXT	    = 0x80,
+	FR_DER_CLASS_PRIVATE	    = 0xC0,
+	FR_DER_CLASS_INVALID	    = 0x04
+} fr_der_tag_class_t;
+
+extern fr_der_tag_constructed_t tag_labels[];
+
+/*
+ *	Create a mapping between FR_TYPE_* and valid FR_DER_TAG_*'s
+ */
+static bool *fr_type_to_der_tags[] = {
+	[FR_TYPE_MAX] = NULL,
+	[FR_TYPE_BOOL] = (bool []){[FR_DER_TAG_BOOLEAN] = true, [FR_DER_TAG_INTEGER] = true, [FR_DER_TAG_NULL] = true, [FR_DER_TAG_MAX] = false},
```

Comment:
maybe put line breaks after the `,` which can make the lines shorter, and a bit easier to read.